File: python/tracing/cohere-traceloop/cohere_traceloop_example.py
#!/usr/bin/env python3
"""
Cohere + Traceloop instrumentation example for Respan tracing.

Uses opentelemetry-instrumentation-cohere (Traceloop) to auto-instrument the
Cohere Python SDK and exports spans to Respan.

Since Cohere uses its own API format (v2/chat), a local HTTP proxy translates
Cohere requests to OpenAI chat completions format for the Respan gateway.
"""
import json
import os
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path

import requests as http_requests
from dotenv import load_dotenv

ROOT_ENV_PATH = Path(__file__).resolve().parents[3] / ".env"
load_dotenv(ROOT_ENV_PATH, override=True)

RESPAN_API_KEY = os.getenv("RESPAN_GATEWAY_API_KEY") or os.getenv("RESPAN_API_KEY")
RESPAN_BASE_URL = (
    os.getenv("RESPAN_GATEWAY_BASE_URL")
    or os.getenv("RESPAN_BASE_URL")
    or "https://api.respan.ai/api"
).rstrip("/")


class CohereToOpenAIProxy(BaseHTTPRequestHandler):
    """Translates Cohere v2/chat requests to OpenAI chat completions."""

    def log_message(self, format, *args):
        pass

    def do_POST(self):
        content_length = int(self.headers.get("Content-Length", 0))
        body = json.loads(self.rfile.read(content_length)) if content_length else {}

        messages = []
        for msg in body.get("messages", []):
            role = msg.get("role", "user")
            content = msg.get("content", "")
            if isinstance(content, list):
                content = " ".join(
                    p.get("text", "") for p in content if isinstance(p, dict) and "text" in p
                )
            if content:
                messages.append({"role": role, "content": content})

        openai_body = {
            "model": "gpt-4o-mini",
            "messages": messages,
            "temperature": body.get("temperature", 0),
            "max_tokens": body.get("max_tokens", 256),
        }

        resp = http_requests.post(
            f"{RESPAN_BASE_URL}/chat/completions",
            json=openai_body,
            headers={
                "Authorization": f"Bearer {RESPAN_API_KEY}",
                "Content-Type": "application/json",
            },
        )

        if resp.status_code != 200:
            self._send_json(resp.status_code, resp.json())
            return

        openai_resp = resp.json()
        choice = openai_resp.get("choices", [{}])[0]
        usage = openai_resp.get("usage", {})

        cohere_resp = {
            "id": openai_resp.get("id", "gen-001"),
            "finish_reason": "COMPLETE",
            "message": {
                "role": "assistant",
                "content": [{"type": "text", "text": choice.get("message", {}).get("content", "")}],
            },
            "usage": {
                "billed_units": {
                    "input_tokens": usage.get("prompt_tokens", 0),
                    "output_tokens": usage.get("completion_tokens", 0),
                },
                "tokens": {
                    "input_tokens": usage.get("prompt_tokens", 0),
                    "output_tokens": usage.get("completion_tokens", 0),
                },
            },
        }

        self._send_json(200, cohere_resp)

    def _send_json(self, status, data):
        payload = json.dumps(data).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def start_proxy():
    server = HTTPServer(("127.0.0.1", 0), CohereToOpenAIProxy)
    port = server.server_address[1]
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server, f"http://127.0.0.1:{port}"


proxy_server, proxy_url = start_proxy()

from opentelemetry.instrumentation.cohere import CohereInstrumentor
from respan_tracing import RespanTelemetry
from respan_tracing.decorators import workflow

telemetry = RespanTelemetry(
    app_name="cohere-traceloop-example",
    api_key=RESPAN_API_KEY,
    is_batching_enabled=False,
    instruments=set(),
)

CohereInstrumentor().instrument()

import cohere

client = cohere.ClientV2(
    api_key="proxy-passthrough",
    base_url=proxy_url,
)


@workflow(name="cohere_traceloop_workflow")
def cohere_workflow(topic: str) -> str:
    response = client.chat(
        model="command-r-plus",
        messages=[
            {"role": "user", "content": f"Write a haiku about {topic}. Output ONLY the haiku."},
        ],
        temperature=0,
        max_tokens=256,
    )
    return response.message.content[0].text


def main() -> None:
    print("=" * 60)
    print("Cohere Traceloop Example")
    print("=" * 60)

    if not RESPAN_API_KEY:
        print("Skipping: RESPAN_API_KEY not set.")
        return

    try:
        result = cohere_workflow(topic="recursion in programming")
        print(result)
    finally:
        telemetry.flush()
        proxy_server.shutdown()


if __name__ == "__main__":
    main()

File: typescript/tracing/cohere-traceloop/cohere_traceloop_example.ts
/**
 * Cohere — Chat completion with Respan tracing via Traceloop.
 *
 * The cohere-ai SDK sends requests to POST /v1/chat with a Cohere-specific
 * body format: { message, model, stream }.
 *
 * Since we route LLM calls through the Respan gateway (OpenAI format),
 * a local HTTP proxy translates the Cohere chat request to OpenAI and back.
 *
 * The @traceloop/instrumentation-cohere patches CohereClient.prototype.chat
 * (and generate, rerank, etc.) to create Traceloop-format OTel spans.
 */

import "dotenv/config";
import http from "node:http";
import https from "node:https";
import { URL } from "node:url";
import { trace } from "@opentelemetry/api";
import { CohereInstrumentation } from "@traceloop/instrumentation-cohere";
import * as CohereModule from "cohere-ai";
import { CohereClient } from "cohere-ai";
import { Respan } from "@respan/respan";

const RESPAN_API_KEY = process.env.RESPAN_API_KEY!;
const RESPAN_BASE_URL = (
  process.env.RESPAN_BASE_URL ?? "https://api.respan.ai/api"
).replace(/\/+$/, "");

// ── 1. Initialize Respan tracing ──────────────────────────────────────
const respan = new Respan({
  apiKey: RESPAN_API_KEY,
  baseURL: RESPAN_BASE_URL,
});
await respan.initialize();

// ── 2. Manually instrument cohere-ai ──────────────────────────────────
const instrumentation = new CohereInstrumentation();
instrumentation.setTracerProvider(trace.getTracerProvider());
instrumentation.manuallyInstrument(CohereModule);

// ── 3. Local proxy: Cohere chat → OpenAI → Respan gateway ────────────
function startProxy(targetBase: string, apiKey: string): Promise<string> {
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      const chunks: Buffer[] = [];
      req.on("data", (c: Buffer) => chunks.push(c));
      req.on("end", () => {
        try {
          const body = JSON.parse(Buffer.concat(chunks).toString());

          // Convert Cohere chat → OpenAI messages
          const messages: { role: string; content: string }[] = [];

          // Add chat history if present
          if (body.chat_history?.length) {
            for (const msg of body.chat_history) {
              messages.push({
                role: msg.role === "CHATBOT" ? "assistant" : "user",
                content: msg.message,
              });
            }
          }

          // Add current message
          messages.push({ role: "user", content: body.message });

          const openaiBody = JSON.stringify({
            model: "gpt-4o-mini",
            messages,
            temperature: body.temperature ?? 0,
            max_tokens: body.max_tokens ?? 256,
          });

          const target = new URL(`${targetBase}/chat/completions`);
          const options: https.RequestOptions = {
            hostname: target.hostname,
            port:
              target.port ||
              (target.protocol === "https:" ? 443 : 80),
            path: target.pathname,
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              "Content-Length": Buffer.byteLength(openaiBody),
              Authorization: `Bearer ${apiKey}`,
            },
          };

          const transport =
            target.protocol === "https:" ? https : http;
          const upstream = transport.request(options, (upRes) => {
            const respChunks: Buffer[] = [];
            upRes.on("data", (c: Buffer) => respChunks.push(c));
            upRes.on("end", () => {
              try {
                const openaiResp = JSON.parse(
                  Buffer.concat(respChunks).toString()
                );
                const choice = openaiResp.choices?.[0];

                // Convert OpenAI response → Cohere format
                const cohereResp = {
                  id: openaiResp.id ?? "cohere-proxy",
                  text: choice?.message?.content ?? "",
                  generation_id: openaiResp.id ?? "gen-proxy",
                  chat_history: [
                    { role: "USER", message: body.message },
                    {
                      role: "CHATBOT",
                      message: choice?.message?.content ?? "",
                    },
                  ],
                  finish_reason: "COMPLETE",
                  meta: {
                    api_version: { version: "1" },
                    tokens: {
                      input_tokens:
                        openaiResp.usage?.prompt_tokens ?? 0,
                      output_tokens:
                        openaiResp.usage?.completion_tokens ?? 0,
                    },
                  },
                };

                const payload = JSON.stringify(cohereResp);
                res.writeHead(200, {
                  "Content-Type": "application/json",
                });
                res.end(payload);
              } catch {
                res.writeHead(502);
                res.end(
                  JSON.stringify({
                    error: "Failed to parse upstream response",
                  })
                );
              }
            });
          });
          upstream.on("error", (e) => {
            res.writeHead(502);
            res.end(JSON.stringify({ error: e.message }));
          });
          upstream.write(openaiBody);
          upstream.end();
        } catch {
          res.writeHead(400);
          res.end(JSON.stringify({ error: "Bad request" }));
        }
      });
    });

    server.listen(0, () => {
      const port = (server.address() as any).port;
      (globalThis as any).__cohereProxy = server;
      resolve(`http://localhost:${port}`);
    });
  });
}

const proxyUrl = await startProxy(RESPAN_BASE_URL, RESPAN_API_KEY);

// ── 4. Create Cohere client pointing at local proxy ───────────────────
const cohere = new CohereClient({
  token: RESPAN_API_KEY,
  environment: proxyUrl,
});

const result = await cohere.chat({
  message: "Write a haiku about recursion in programming.",
  model: "command-r-plus",
});

console.log(result.text);

// ── 5. Flush and cleanup ──────────────────────────────────────────────
await respan.flush();
((globalThis as any).__cohereProxy as http.Server).close();
